File: litho-core/src/main/java/com/facebook/litho/DynamicPropsManager.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.Pair;
import android.util.SparseArray;
import android.view.View;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Takes care of dynamic Props
 *
 * <p>Usage from {@link DynamicPropsExtension}
 *
 * <p>Unique per {@link DynamicPropsExtension} instance, which calls {@link
 * #onBindComponentToContent(Component, ComponentContext, SparseArray, Object)} and {#link {@link
 * #onUnbindComponent(Component, SparseArray, Object)}} when it binds and unbinds Components
 * respectively.
 *
 * <p>When Component is being bound, a DynamicPropsManager subscribes to {@link DynamicValue}s that
 * its represent dynamic props, and keeps the mounted content the Components is rendered into in
 * sync with {@link DynamicValue#get()} until the Component gets unbound, at which point, the
 * DynamicPropsManager unsubscribes from the DynamicValues.
 */
public class DynamicPropsManager implements DynamicValue.OnValueChangeListener {
  public static final int KEY_ALPHA = 1;
  public static final int KEY_TRANSLATION_X = 2;
  public static final int KEY_TRANSLATION_Y = 3;
  public static final int KEY_SCALE_X = 4;
  public static final int KEY_SCALE_Y = 5;
  public static final int KEY_ELEVATION = 6;
  public static final int KEY_BACKGROUND_COLOR = 7;
  public static final int KEY_ROTATION = 8;
  public static final int KEY_BACKGROUND_DRAWABLE = 9;
  public static final int KEY_FOREGROUND_COLOR = 10;

  private static final DynamicValue<?>[] sEmptyArray = new DynamicValue[0];

  private final Map<DynamicValue<?>, Set<Pair<Component, SparseArray<? extends DynamicValue<?>>>>>
      mDependentComponents = new HashMap<>();
  private final Map<Component, Set<DynamicValue<?>>> mAffectingDynamicValues = new HashMap<>();
  private final Map<Component, Object> mContents = new HashMap<>();

  void onBindComponentToContent(
      final Component component,
      final @Nullable ComponentContext scopedContext,
      final @Nullable SparseArray<? extends DynamicValue<?>> commonDynamicProps,
      final Object content) {
    final boolean hasCommonDynamicPropsToBind =
        hasCommonDynamicPropsToBind(commonDynamicProps, content);

    if (!hasCommonDynamicPropsToBind && !hasCustomDynamicProps(component)) {
      return;
    }

    final Set<DynamicValue<?>> dynamicValues = new HashSet<>();

    if (commonDynamicProps != null && hasCommonDynamicPropsToBind) {
      // Go through all common dynamic props
      for (int i = 0; i < commonDynamicProps.size(); i++) {
        final int key = commonDynamicProps.keyAt(i);
        final DynamicValue<?> value = commonDynamicProps.valueAt(i);

        bindCommonDynamicProp(key, value, (View) content);

        addDependentComponentAndSubscribeIfNeeded(value, new Pair<>(component, commonDynamicProps));
        dynamicValues.add(value);
      }
    }

    final DynamicValue<?>[] customDynamicProps = getCustomDynamicProps(component);
    // Go through all the custom dynamic props
    for (int i = 0; i < customDynamicProps.length; i++) {
      final @Nullable DynamicValue<?> value = customDynamicProps[i];

      try {
        ((SpecGeneratedComponent) component)
            .bindDynamicProp(i, value != null ? value.get() : null, content);

        addDependentComponentAndSubscribeIfNeeded(value, new Pair<>(component, commonDynamicProps));
        dynamicValues.add(value);
      } catch (Exception e) {
        if (scopedContext != null) {
          ComponentUtils.handle(scopedContext, e);
        } else {
          ComponentUtils.rethrow(e);
        }
      }
    }

    mAffectingDynamicValues.put(component, dynamicValues);
    mContents.put(component, content);
  }

  void onUnbindComponent(
      Component component,
      final @Nullable SparseArray<? extends DynamicValue<?>> commonDynamicProps,
      Object content) {
    if (!hasCommonDynamicPropsToBind(commonDynamicProps, content)
        && !hasCustomDynamicProps(component)) {
      return;
    }

    mContents.remove(component);

    final Set<DynamicValue<?>> dynamicValues = mAffectingDynamicValues.get(component);
    if (dynamicValues == null) {
      return;
    }

    for (DynamicValue<?> value : dynamicValues) {
      removeDependentComponentAndUnsubscribeIfNeeded(
          value, new Pair<>(component, commonDynamicProps));
    }

    // Go through all common dynamic props to reset them if they were set
    if (commonDynamicProps != null) {
      for (int i = 0; i < commonDynamicProps.size(); i++) {
        final int key = commonDynamicProps.keyAt(i);
        resetDynamicValues(key, content);
      }
    }

    mAffectingDynamicValues.remove(component);
  }

  private static void resetDynamicValues(int key, Object content) {
    if (!(content instanceof View)) {
      return;
    }

    final View target = (View) content;
    switch (key) {
      case KEY_ALPHA:
        if (target.getAlpha() != 1) {
          target.setAlpha(1);
        }
        break;

      case KEY_TRANSLATION_X:
        if (target.getTranslationX() != 0) {
          target.setTranslationX(0);
        }
        break;

      case KEY_TRANSLATION_Y:
        if (target.getTranslationY() != 0) {
          target.setTranslationY(0);
        }
        break;

      case KEY_SCALE_X:
        if (target.getScaleX() != 1) {
          target.setScaleX(1);
        }
        break;

      case KEY_SCALE_Y:
        if (target.getScaleY() != 1) {
          target.setScaleY(1);
        }
        break;

      case KEY_ELEVATION:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && target.getElevation() != 0) {
          target.setElevation(0);
        }
        break;

      case KEY_ROTATION:
        if (target.getRotation() != 0) {
          target.setRotation(0);
        }
        break;

      case KEY_BACKGROUND_COLOR:
      case KEY_BACKGROUND_DRAWABLE:
        if (target.getBackground() != null) {
          target.setBackground(null);
        }
        break;

      case KEY_FOREGROUND_COLOR:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && target.getForeground() != null) {
          ViewUtils.setViewForeground(target, null);
        }
    }
  }

  private void addDependentComponentAndSubscribeIfNeeded(
      @Nullable DynamicValue<?> value,
      Pair<Component, SparseArray<? extends DynamicValue<?>>> componentWithProps) {
    if (value == null) {
      return;
    }
    Set<Pair<Component, SparseArray<? extends DynamicValue<?>>>> dependentComponents =
        mDependentComponents.get(value);

    if (dependentComponents == null) {
      dependentComponents = new HashSet<>();
      mDependentComponents.put(value, dependentComponents);

      value.attachListener(this);
    }

    dependentComponents.add(componentWithProps);
  }

  private void removeDependentComponentAndUnsubscribeIfNeeded(
      @Nullable DynamicValue<?> value,
      Pair<Component, SparseArray<? extends DynamicValue<?>>> componentWithProps) {
    if (value == null) {
      return;
    }
    final Set<Pair<Component, SparseArray<? extends DynamicValue<?>>>> dependentComponents =
        mDependentComponents.get(value);
    if (dependentComponents == null) {
      return;
    }

    dependentComponents.remove(componentWithProps);

    if (dependentComponents.isEmpty()) {
      mDependentComponents.remove(value);
      value.detach(this);
    }
  }

  private static void bindCommonDynamicProp(int key, @Nullable DynamicValue<?> value, View target) {
    switch (key) {
      case KEY_ALPHA:
        target.setAlpha(DynamicPropsManager.<Float>resolve(value));
        break;

      case KEY_TRANSLATION_X:
        target.setTranslationX(DynamicPropsManager.<Float>resolve(value));
        break;

      case KEY_TRANSLATION_Y:
        target.setTranslationY(DynamicPropsManager.<Float>resolve(value));
        break;

      case KEY_SCALE_X:
        target.setScaleX(DynamicPropsManager.<Float>resolve(value));
        break;

      case KEY_SCALE_Y:
        target.setScaleY(DynamicPropsManager.<Float>resolve(value));
        break;

      case KEY_ELEVATION:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
          target.setElevation(DynamicPropsManager.<Float>resolve(value));
        }
        break;

      case KEY_BACKGROUND_COLOR:
        target.setBackgroundColor(DynamicPropsManager.<Integer>resolve(value));
        break;

      case KEY_ROTATION:
        target.setRotation(DynamicPropsManager.<Float>resolve(value));
        break;

      case KEY_BACKGROUND_DRAWABLE:
        target.setBackground(DynamicPropsManager.<Drawable>resolve(value));
        break;

      case KEY_FOREGROUND_COLOR:
        ViewUtils.setViewForeground(target, DynamicPropsManager.<Integer>resolve(value));
    }
  }

  @SuppressWarnings("unchecked")
  private static <T> T resolve(DynamicValue<?> dynamicValue) {
    return (T) dynamicValue.get();
  }

  @Override
  public void onValueChange(DynamicValue value) {
    final Set<Pair<Component, SparseArray<? extends DynamicValue<?>>>> dependentComponents =
        mDependentComponents.get(value);
    if (dependentComponents == null) {
      return;
    }

    // It's possible that applying a dynamic prop could bind or unbind a component - snapshot the
    // components here to prevent a ConcurrentModificationException during iteration
    Pair<Component, SparseArray<DynamicValue<?>>>[] dependentComponentSnapshot =
        dependentComponents.toArray(new Pair[dependentComponents.size()]);
    for (Pair<Component, SparseArray<DynamicValue<?>>> componentWithProps :
        dependentComponentSnapshot) {
      final Component component = componentWithProps.first;
      final SparseArray<DynamicValue<?>> commonDynamicProps = componentWithProps.second;
      final Object content = mContents.get(component);
      if (content == null) {
        continue;
      }

      if (hasCommonDynamicPropsToBind(commonDynamicProps, content)) {
        for (int i = 0; i < commonDynamicProps.size(); i++) {
          if (commonDynamicProps.valueAt(i) == value) {
            bindCommonDynamicProp(commonDynamicProps.keyAt(i), value, (View) content);
          }
        }
      }

      final DynamicValue<?>[] dynamicProps = getCustomDynamicProps(component);
      for (int i = 0; i < dynamicProps.length; i++) {
        if (value == dynamicProps[i]) {
          ((SpecGeneratedComponent) component).bindDynamicProp(i, value.get(), content);
        }
      }
    }
  }

  /**
   * Common dynamic props could only be bound to Views. To make it work for the LayoutSpec and
   * MountDrawableSpec components we create a wrapping HostComponent and copy the dynamic props
   * there. Thus DynamicPropsManager should ignore non-MountViewSpecs
   *
   * @param commonDynamicProps to consider
   * @return true if Component has common dynamic props, that DynamicPropsManager should take an
   *     action on
   */
  private static boolean hasCommonDynamicPropsToBind(
      final @Nullable SparseArray<? extends DynamicValue<?>> commonDynamicProps, Object content) {
    return CollectionsUtils.isNotNullOrEmpty(commonDynamicProps) && content instanceof View;
  }

  private static boolean hasCustomDynamicProps(Component component) {
    return component instanceof SpecGeneratedComponent
        && ((SpecGeneratedComponent) component).getDynamicProps().length > 0;
  }

  private static DynamicValue<?>[] getCustomDynamicProps(Component component) {
    return component instanceof SpecGeneratedComponent
        ? ((SpecGeneratedComponent) component).getDynamicProps()
        : sEmptyArray;
  }

  @VisibleForTesting
  boolean hasCachedContent(Component component) {
    return mContents.containsKey(component);
  }
}

File: litho-core/src/main/java/com/facebook/litho/DynamicPropsManager.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.os.Build
import android.util.Pair
import android.util.SparseArray
import android.view.View
import androidx.annotation.VisibleForTesting
import com.facebook.litho.CollectionsUtils.isNotNullOrEmpty
import com.facebook.litho.ComponentUtils.handle
import com.facebook.litho.ComponentUtils.rethrow

/**
 * Takes care of dynamic Props
 *
 * Usage from [DynamicPropsExtension]
 *
 * Unique per [DynamicPropsExtension] instance, which calls [onBindComponentToContent] and
 * [onUnbindComponent] when it binds and unbinds Components respectively.
 *
 * When Component is being bound, a DynamicPropsManager subscribes to [DynamicValue]s that its
 * represent dynamic props, and keeps the mounted content the Components is rendered into in sync
 * with [DynamicValue.get] until the Component gets unbound, at which point, the DynamicPropsManager
 * unsubscribes from the DynamicValues.
 */
class DynamicPropsManager : DynamicValue.OnValueChangeListener<Any?> {
  private val dependentComponents:
      MutableMap<
          DynamicValue<Any?>, MutableSet<Pair<Component, SparseArray<out DynamicValue<Any?>>>>> =
      HashMap()
  private val affectingDynamicValues: MutableMap<Component, Set<DynamicValue<Any?>?>> = HashMap()
  private val contents: MutableMap<Component, Any> = HashMap()

  fun onBindComponentToContent(
      component: Component,
      scopedContext: ComponentContext?,
      commonDynamicProps: SparseArray<out DynamicValue<Any?>>?,
      content: Any
  ) {
    val hasCommonDynamicPropsToBind = hasCommonDynamicPropsToBind(commonDynamicProps, content)
    if (!hasCommonDynamicPropsToBind && !hasCustomDynamicProps(component)) {
      return
    }
    val dynamicValues: MutableSet<DynamicValue<Any?>?> = HashSet()
    if (commonDynamicProps != null && hasCommonDynamicPropsToBind) {
      // Go through all common dynamic props
      for (i in 0 until commonDynamicProps.size()) {
        val key = commonDynamicProps.keyAt(i)
        val value = commonDynamicProps.valueAt(i)
        bindCommonDynamicProp(key, value, content as View)
        addDependentComponentAndSubscribeIfNeeded(value, Pair(component, commonDynamicProps))
        dynamicValues.add(value)
      }
    }
    val customDynamicProps = getCustomDynamicProps(component)
    // Go through all the custom dynamic props
    for (i in customDynamicProps.indices) {
      val value = customDynamicProps[i]
      try {
        (component as SpecGeneratedComponent).bindDynamicProp(i, value?.get(), content)
        addDependentComponentAndSubscribeIfNeeded(value, Pair(component, commonDynamicProps))
        dynamicValues.add(value)
      } catch (e: Exception) {
        if (scopedContext != null) {
          handle(scopedContext, e)
        } else {
          rethrow(e)
        }
      }
    }
    affectingDynamicValues[component] = dynamicValues
    contents[component] = content
  }

  fun onUnbindComponent(
      component: Component,
      commonDynamicProps: SparseArray<out DynamicValue<Any?>>?,
      content: Any
  ) {
    if (!hasCommonDynamicPropsToBind(commonDynamicProps, content) &&
        !hasCustomDynamicProps(component)) {
      return
    }
    contents.remove(component)
    val dynamicValues = affectingDynamicValues[component] ?: return
    for (value in dynamicValues) {
      removeDependentComponentAndUnsubscribeIfNeeded(value, Pair(component, commonDynamicProps))
    }

    // Go through all common dynamic props to reset them if they were set
    if (commonDynamicProps != null) {
      for (i in 0 until commonDynamicProps.size()) {
        val key = commonDynamicProps.keyAt(i)
        resetDynamicValues(key, content)
      }
    }
    affectingDynamicValues.remove(component)
  }

  private fun addDependentComponentAndSubscribeIfNeeded(
      value: DynamicValue<Any?>?,
      componentWithProps: Pair<Component, SparseArray<out DynamicValue<Any?>>>
  ) {
    if (value == null) {
      return
    }
    var dependentComponents = dependentComponents[value]
    if (dependentComponents == null) {
      dependentComponents = HashSet()
      this.dependentComponents[value] = dependentComponents
      value.attachListener(this)
    }
    dependentComponents.add(componentWithProps)
  }

  private fun removeDependentComponentAndUnsubscribeIfNeeded(
      value: DynamicValue<Any?>?,
      componentWithProps: Pair<Component, SparseArray<out DynamicValue<Any?>>>
  ) {
    if (value == null) {
      return
    }
    val dependentComponents = dependentComponents[value] ?: return
    dependentComponents.remove(componentWithProps)
    if (dependentComponents.isEmpty()) {
      this.dependentComponents.remove(value)
      value.detach(this)
    }
  }

  override fun onValueChange(value: DynamicValue<Any?>) {
    val dependentComponents = dependentComponents[value] ?: return

    // It's possible that applying a dynamic prop could bind or unbind a component - snapshot the
    // components here to prevent a ConcurrentModificationException during iteration
    val dependentComponentSnapshot: Array<Pair<Component, SparseArray<out DynamicValue<Any?>>>> =
        dependentComponents.toTypedArray()
    for (componentWithProps in dependentComponentSnapshot) {
      val component = componentWithProps.first
      val commonDynamicProps = componentWithProps.second
      val content = contents[component] ?: continue
      if (hasCommonDynamicPropsToBind(commonDynamicProps, content)) {
        for (i in 0 until commonDynamicProps.size()) {
          if (commonDynamicProps.valueAt(i) === value) {
            bindCommonDynamicProp(commonDynamicProps.keyAt(i), value, content as View)
          }
        }
      }
      val dynamicProps = getCustomDynamicProps(component)
      for (i in dynamicProps.indices) {
        if (value === dynamicProps[i]) {
          (component as SpecGeneratedComponent).bindDynamicProp(i, value.get(), content)
        }
      }
    }
  }

  @VisibleForTesting
  fun hasCachedContent(component: Component): Boolean {
    return contents.containsKey(component)
  }

  companion object {
    const val KEY_ALPHA: Int = 1
    const val KEY_TRANSLATION_X: Int = 2
    const val KEY_TRANSLATION_Y: Int = 3
    const val KEY_TRANSLATION_Z: Int = 4
    const val KEY_SCALE_X: Int = 5
    const val KEY_SCALE_Y: Int = 6
    const val KEY_ELEVATION: Int = 7
    const val KEY_BACKGROUND_COLOR: Int = 8
    const val KEY_ROTATION: Int = 9
    const val KEY_ROTATION_X: Int = 10
    const val KEY_ROTATION_Y: Int = 11
    const val KEY_BACKGROUND_DRAWABLE: Int = 12
    const val KEY_FOREGROUND_COLOR: Int = 13

    private fun resetDynamicValues(key: Int, content: Any) {
      if (content !is View) {
        return
      }
      when (key) {
        KEY_ALPHA ->
            if (content.alpha != 1f) {
              content.alpha = 1f
            }
        KEY_TRANSLATION_X ->
            if (content.translationX != 0f) {
              content.translationX = 0f
            }
        KEY_TRANSLATION_Y ->
            if (content.translationY != 0f) {
              content.translationY = 0f
            }
        KEY_TRANSLATION_Z ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
              if (content.translationZ != 0f) {
                content.translationZ = 0f
              }
            }
        KEY_SCALE_X ->
            if (content.scaleX != 1f) {
              content.scaleX = 1f
            }
        KEY_SCALE_Y ->
            if (content.scaleY != 1f) {
              content.scaleY = 1f
            }
        KEY_ELEVATION ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && content.elevation != 0f) {
              content.elevation = 0f
            }
        KEY_ROTATION ->
            if (content.rotation != 0f) {
              content.rotation = 0f
            }
        KEY_ROTATION_X ->
            if (content.rotationX != 0f) {
              content.rotationX = 0f
            }
        KEY_ROTATION_Y ->
            if (content.rotationY != 0f) {
              content.rotationY = 0f
            }
        KEY_BACKGROUND_COLOR,
        KEY_BACKGROUND_DRAWABLE ->
            if (content.background != null) {
              content.background = null
            }
        KEY_FOREGROUND_COLOR ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && content.foreground != null) {
              setViewForeground(content, null)
            }
      }
    }

    private fun bindCommonDynamicProp(key: Int, value: DynamicValue<*>, target: View) {
      when (key) {
        KEY_ALPHA -> target.alpha = resolve(value)
        KEY_TRANSLATION_X -> target.translationX = resolve(value)
        KEY_TRANSLATION_Y -> target.translationY = resolve(value)
        KEY_TRANSLATION_Z ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
              target.translationZ = resolve(value)
            }
        KEY_SCALE_X -> target.scaleX = resolve(value)
        KEY_SCALE_Y -> target.scaleY = resolve(value)
        KEY_ELEVATION ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
              target.elevation = resolve(value)
            }
        KEY_BACKGROUND_COLOR -> target.setBackgroundColor(resolve(value))
        KEY_ROTATION -> target.rotation = resolve(value)
        KEY_ROTATION_X -> target.rotationX = resolve(value)
        KEY_ROTATION_Y -> target.rotationY = resolve(value)
        KEY_BACKGROUND_DRAWABLE -> target.background = resolve(value)
        KEY_FOREGROUND_COLOR -> setViewForeground(target, resolve<Int>(value))
      }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> resolve(dynamicValue: DynamicValue<*>): T {
      return dynamicValue.get() as T
    }

    /**
     * Common dynamic props could only be bound to Views. To make it work for the LayoutSpec and
     * MountDrawableSpec components we create a wrapping HostComponent and copy the dynamic props
     * there. Thus DynamicPropsManager should ignore non-MountViewSpecs
     *
     * @param commonDynamicProps to consider
     * @return true if Component has common dynamic props, that DynamicPropsManager should take an
     *   action on
     */
    private fun hasCommonDynamicPropsToBind(
        commonDynamicProps: SparseArray<out DynamicValue<*>>?,
        content: Any
    ): Boolean {
      return isNotNullOrEmpty(commonDynamicProps) && content is View
    }

    private fun hasCustomDynamicProps(component: Component): Boolean {
      return (component is SpecGeneratedComponent && component.dynamicProps.isNotEmpty())
    }

    @Suppress("UNCHECKED_CAST")
    private fun getCustomDynamicProps(component: Component): Array<DynamicValue<Any?>?> {
      return if (component is SpecGeneratedComponent)
          component.dynamicProps as Array<DynamicValue<Any?>?>
      else emptyArray()
    }
  }
}
